### ISSUE TITLE
Add dflash compatible training step

### ISSUE THREAD

**Issue by dsikka** [vllm-project/speculators#252]
Dflash runs a single forward step which produces a tensor of shape $[vocab\\_size,L]$  This is in contrast to the eagle3 train time test paradigm which runs the drafter forward for $T$ steps.  The drafter produces a block of tokens of length $B$ for each prefix (total of $L/B$ prefixes).  This block structure is entirely accomplished using the attention mask. We calculate the loss using the same function as eagle 3.  We then apply the loss mask to this and calculate gradients on the model.  

See further details: [[RFC][Training] Add dflash support in training #248](https://github.com/vllm-project/speculators/issues/248)
